torch/utils/data/datapipes/iter/selecting.py

```diff
@@ -12,28 +12,54 @@ class FilterIterDataPipe(MapIterDataPipe):
 
     Iterable DataPipe to filter elements from datapipe according to filter_fn.
     args:
-        datapipe: Iterable DataPipe being filterd
+        datapipe: Iterable DataPipe being filtered
         filter_fn: Customized function mapping an element to a boolean.
         fn_args: Positional arguments for `filter_fn`
         fn_kwargs: Keyword arguments for `filter_fn`
+        drop_empty_batches: By default, drops batch if it is empty after filtering instead of keeping an empty list
+        nesting_level: Determines which level the fn gets applied to, by default it applies to the top level (= 0)
     """
     def __init__(self,
                  datapipe: IterDataPipe[T_co],
                  filter_fn: Callable[..., bool],
                  fn_args: Optional[Tuple] = None,
                  fn_kwargs: Optional[Dict] = None,
+                 drop_empty_batches: bool = True,
+                 nesting_level: int = 0,
                  ) -> None:
-        super().__init__(datapipe, fn=filter_fn, fn_args=fn_args, fn_kwargs=fn_kwargs)
+        self.drop_empty_batches = drop_empty_batches
+        super().__init__(datapipe, fn=filter_fn, fn_args=fn_args, fn_kwargs=fn_kwargs, nesting_level=nesting_level)
+
+    def _merge(self, data, mask):
+        result = []
+        for i, b in zip(data, mask):
+            if isinstance(b, list):
+                t = self._merge(i, b)
+                if len(t) > 0 or not self.drop_empty_batches:
+                    result.append(t)
+            else:
+                if not isinstance(b, bool):
+                    raise ValueError("Boolean output is required for "
+                                     "`filter_fn` of FilterIterDataPipe")
+                if b:
+                    result.append(i)
+        return result
 
     def __iter__(self) -> Iterator[T_co]:
         res: bool
         for data in self.datapipe:
-            res = self.fn(data, *self.args, **self.kwargs)
-            if not isinstance(res, bool):
-                raise ValueError("Boolean output is required for "
-                                 "`filter_fn` of FilterIterDataPipe")
-            if res:
-                yield data
+            if self.nesting_level == 0:
```

Comment:
Why is this conditional necessary `_apply` tests if `nesting_level` is zero too:

```
    def _apply(self, data, nesting_level):
        if nesting_level == 0:
            return self.fn(data, *self.args, **self.kwargs)
        elif nesting_level > 0:
            if not isinstance(data, list):
                raise IndexError(f"nesting_level {self.nesting_level} out of range (exceeds data pipe depth)")
            result = [self._apply(i, nesting_level - 1) for i in data]
            return result
        else:
            if isinstance(data, list):
                result = [self._apply(i, nesting_level) for i in data]
                return result
            else:
```

Reply:
Is it because you want to return a non-list when `nesting_level == 0`, but a list otherwise? But it seems `_merge` ought to handle that?

Reply:
This condition is within `for data in self.datapipe` and we need it for the case where the `datapipe` is a 1D array (e.g. [0,1,2,3]) and `nest_level = 0`. In this case, `data` is a non-list, if we skip the condition and go into the else statement:
```
mask = self._apply(data, self.nesting_level)
merged = self._merge(data, mask)
```
`mask` will equal to a `bool`, instead of a list of `bool`, and `_merge` doesn't handle the case where `mask` is not a list of bool because it calls `zip(data, mask)`.

Let me know if this makes sense or if there is a cleaner implementation. One way may be to check if `data` and `mask` are non-list within `_merge` instead.

Reply:
Yeah, don't do merging based on the types of the inputs, instead, pass in `self.nesting_level`. That will prevent type confusion when you get a list but actually you're supposed to stop processing the nesting.

Reply:
Just to recap, the current implementation takes advantage of the fact that `_apply` takes in `self.nesting_level` and has the logic to figure out the right level to apply the filter. Therefore, when `_merge` is subsequently called, `.merge` does not need `self.nesting_level` and only needs to traverse to where the booleans (non-list) are based on the inputs. I believe if we choose to pass in `self.nesting_level` to `_merge`, we will be basically re-doing basically what `_apply` already does for us.

If that implementation is cleaner, then we should combine what `_apply` and `_merge` do into one step. Here is my implementation of that into a new function `_applyFilter`. Let me know what you think.

https://github.com/pytorch/pytorch/blob/DataPipesAPIClarity/torch/utils/data/datapipes/iter/selecting.py

Reply:
Well, regardless of whether or not the code is cleaner, the new version is more efficient because you avoid having to allocate the intermediate list structure twice (_merge).

I'm guessing that you split the implementation into _apply and _merge to make it so that the functions did less individually. But I agree that the functions still do very much similar things, and you only use these helpers once. You might consider splitting things out again later when you have extra call sites, but a smaller, more compact piece of code is better to start with, I guess.

Reply:
Okay - I will create a new PR with all the changes I implemented based on your feedback. Thanks!